File: python/src/bursar/credits/postgres/repositories/bucket.py
from __future__ import annotations

from decimal import Decimal

from bursar.credits.postgres.repositories._types import DbQuery
from bursar.credits.postgres.repositories._utils import (
    require_mapping_row,
    validate_non_empty,
    validate_row,
)
from bursar.credits.postgres.repositories.schemas import BucketBalanceRow, BucketEnvelopeRow, SweepRow


class BucketRepository:
    """Repository for credit bucket and expiry operations.

    All methods call Postgres RPCs via the callproc function.
    Returns None when the RPC returns no rows (no result).
    Returns typed Pydantic models for successful results.
    """

    def __init__(self, callproc: DbQuery) -> None:
        self._callproc = callproc

    def get_bucket_balances(self, user_id: str) -> BucketEnvelopeRow:
        """Get all credit bucket balances for a user.

        Args:
            user_id: The user ID.

        Returns:
            BucketEnvelopeRow if found, None if the user has no buckets.
        """
        validate_non_empty(user_id, "user_id")
        rows = self._callproc("get_credit_bucket_balances", [user_id])
        buckets = [validate_row(BucketBalanceRow, row, "BucketRepository.get_bucket_balances") for row in rows or []]
        total = sum(
            (Decimal(str(row.balance)) for row in buckets),
            Decimal(0),
        )
        return validate_row(
            BucketEnvelopeRow,
            {
                "user_id": user_id,
                "buckets": buckets,
                "total_balance": total,
            },
            "BucketRepository.get_bucket_balances",
        )

    def sweep_expired_credits(
        self,
        dry_run: bool = False,
        user_id: str | None = None,
        limit: int = 100,
    ) -> SweepRow:
        """Expire at most ``limit`` eligible credit lots."""
        if limit < 1:
            raise ValueError("limit must be positive")
        rows = self._callproc("sweep_expired_lots", [limit, user_id, dry_run])
        row = require_mapping_row(rows, "BucketRepository.sweep_expired_credits")
        return validate_row(
            SweepRow,
            row,
            "BucketRepository.sweep_expired_credits",
            indeterminate=not dry_run,
        )

File: javascript/src/credits/postgres/repositories/bucket.ts
import { z } from "zod";
import { Decimal } from "decimal.js";
import type { CallProc } from "../../../shared/postgres-types.js";
import { postgresUuid, requireRow, safeParse } from "../../../shared/postgres-validation.js";

const decimal = z.union([z.string().min(1), z.number().finite()] as const);

const BucketRowSchema = z
  .object({
    bucket_key: z.string().min(1),
    label: z.string().min(1),
    priority: z.number().int().nonnegative(),
    expires: z.boolean(),
    balance: decimal,
  })
  .strict();

const BucketEnvelopeRowSchema = z
  .object({
    user_id: postgresUuid,
    buckets: z.array(BucketRowSchema),
    total_balance: decimal,
  })
  .strict();

const SweepRowSchema = z
  .object({
    expired_count: z.number().int().nonnegative(),
    expired_amount: decimal,
    expired_by_bucket: z.record(z.string().min(1), decimal),
  })
  .strict();

export type BucketEnvelopeRow = z.infer<typeof BucketEnvelopeRowSchema>;
export type SweepRow = z.infer<typeof SweepRowSchema>;

/** Repository for credit bucket operations. */
export class BucketRepository {
  constructor(private callproc: CallProc) {}

  /** Fetch per-bucket credit balances for a user. */
  async getBucketBalances(userId: string): Promise<BucketEnvelopeRow> {
    const rows = await this.callproc("get_credit_bucket_balances", [userId]);
    const buckets = (rows ?? []).map((row) =>
      safeParse(BucketRowSchema, row, "BucketRepository.getBucketBalances"),
    );
    const totalBalance = buckets
      .reduce((total, row) => total.plus(row.balance), new Decimal(0))
      .toString();
    return safeParse(
      BucketEnvelopeRowSchema,
      {
        user_id: userId,
        buckets,
        total_balance: totalBalance,
      },
      "BucketRepository.getBucketBalances",
    );
  }

  /** Sweep expired credit grants. */
  async sweepExpiredCredits(dryRun = false, userId?: string, limit = 100): Promise<SweepRow> {
    const rows = await this.callproc("sweep_expired_lots", [limit, userId ?? null, dryRun]);
    return safeParse(
      SweepRowSchema,
      requireRow(rows, "BucketRepository.sweepExpiredCredits"),
      "BucketRepository.sweepExpiredCredits",
    );
  }
}
